go-controller/pkg/ovn/egressgw.go

```diff
@@ -39,6 +41,114 @@ type ovnRoute struct {
 	shouldExist bool
 }
 
+type externalRouteInfo struct {
+	sync.RWMutex
+	// podExternalRoutes is a cache keeping the LR routes added to the GRs when
+	// the k8s.ovn.org/routing-external-gws annotation is used. The first map key
+	// is the podIP, the second the GW and the third the GR
+	routes map[string]map[string]string
+}
+
+func (oc *Controller) ensureRouteInfoLocked(podName ktypes.NamespacedName, readOnly bool) (*externalRouteInfo, func(), error) {
+	oc.egressGWCacheMutex.Lock()
+	routeInfo := oc.egressGWCache[podName]
+	routeInfoExisted := false
+	if routeInfo == nil {
+		routeInfo = &externalRouteInfo{
+			routes: make(map[string]map[string]string),
+		}
+		// we are creating routeInfo and going to set it in podExternalRoutes map
+		// so safe to hold the lock while we create and add it
+		defer oc.egressGWCacheMutex.Unlock()
+		oc.egressGWCache[podName] = routeInfo
+	} else {
+		routeInfoExisted = true
+		// if we found and existing nsInfo, do not hold the namespaces lock
+		// while waiting for nsInfo to Lock
+		oc.egressGWCacheMutex.Unlock()
+	}
+
+	var unlockFunc func()
+	if readOnly {
+		unlockFunc = routeInfo.RUnlock
+		routeInfo.RLock()
+	} else {
+		unlockFunc = routeInfo.Unlock
+		routeInfo.Lock()
+	}
+
+	if routeInfoExisted {
+		// Check that the namespace wasn't deleted while we were waiting for the lock
+		oc.egressGWCacheMutex.Lock()
+		defer oc.egressGWCacheMutex.Unlock()
+		if routeInfo != oc.egressGWCache[podName] {
+			unlockFunc()
+			return nil, nil, fmt.Errorf("pod %s, was removed during ensure route info", podName)
+		}
+	}
+
+	return routeInfo, unlockFunc, nil
+}
+
+func (oc *Controller) getRouteInfosForGateway(gatewayIP, namespace string) []*externalRouteInfo {
+	oc.egressGWCacheMutex.RLock()
+	defer oc.egressGWCacheMutex.RUnlock()
+
+	routes := make([]*externalRouteInfo, 0)
+	for namespacedName, routeInfo := range oc.egressGWCache {
+		if namespacedName.Namespace != namespace {
+			continue
+		}
+
+		for _, route := range routeInfo.routes {
+			if _, ok := route[gatewayIP]; ok {
+				routeInfo.Lock()
+				routes = append(routes, routeInfo)
+			}
+		}
+	}
+
+	return routes
+}
+
+func (oc *Controller) getRouteInfosForNamespace(namespace string) []*externalRouteInfo {
+	oc.egressGWCacheMutex.RLock()
+	defer oc.egressGWCacheMutex.RUnlock()
+
+	routes := make([]*externalRouteInfo, 0)
+	for namespacedName, routeInfo := range oc.egressGWCache {
+		if namespacedName.Namespace != namespace {
+			continue
+		}
+		routeInfo.Lock()
+		routes = append(routes, routeInfo)
+	}
+
+	return routes
+}
+
+func (oc *Controller) deleteRouteInfoLocked(name ktypes.NamespacedName) *externalRouteInfo {
```

Comment:
@trozet can you add some code comments here explaining the locking? eg we don't want to block trying to get routeInfo.Lock() while holding the egressGWCacheMutex if we can help it.

Reply:
done